### DISCUSSION TITLE
How can I get node embedding from the basic GCN encoder?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#6542] [Q&A] [answered]
I think that this question is duplicated with this question (https://github.com/pyg-team/pytorch_geometric/discussions/4561), but I need more detail answer for the problem.
(Fortunately, I solve the previous problem by running the code in other server with ubuntu version 18.)

First, I prepared the PyG Data from networkx graph and got the dataset as below:
```
Data(edge_index=[2, 2475552], weight=[2475552], x=[1986, 1648])
````
The original networkx data was made from two csv files, one is adjacency matrix (row: node, column: node) and another is feature matrix (row: node, column: 1,648 of features).
Note that I don't have edge attributes in the graph.

Then, I ran the basic GCN using `GCNConv` in  `torch_geometric.nn`.
```
import torch
import torch.nn.functional as F
from torch_geometric.nn import GCNConv

feature_dim = pyg_graph.num_node_features

class GCNEncoder(torch.nn.Module):
    def __init__(self):
        super(GCNEncoder, self).__init__()
        self.conv1 = GCNConv(feature_dim, 256)
        self.conv2 = GCNConv(256, 16)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index

        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = self.conv2(x, edge_index)

        return x
```
Because I only want the node embeddings with 16 dimensions, I did not add other activation functions.

By using `GCNEncoder`, I tried run the model directly input the dataset, but I faced the error.
```
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu") 
print("The model will be running on", device, "device\n") 

ppi_model = GCNEncoder().to(device)
ppi_data = ppi_pyg_graph.to(device)

ppi_out = ppi_model(ppi_data)
```
`RuntimeError: "addmm_cuda" not implemented for 'Long'`

As in https://github.com/pyg-team/pytorch_geometric/discussions/4561, I tried to add edge attributes, but I don't know where should I fix.
Can anyone suggest idea for this problem?
Thanks!

**Comment by rusty1s** [comment, marked as answer]
Do you have a minimal example to reproduce? What `dtype` does `edge_index`, `edge_weight` and `x` have? What does `data.validate()` report?

**Reply by songsong0425** [reply]
I hope this reply be helpful to you.
I run the below code to generate the PyG Data. If there are any errors, please notify to me.
```
import networkx as nx
import numpy as np
import pandas as pd
import torch
from torch_geometric.utils.convert import from_networkx
from torch_geometric.data import Data

ppi_feature=pd.read_csv('PPI_Features_Identity.csv')
ppi_edges=pd.read_csv('PPI_Adjacency.csv', index_col=0)

ppi_feature = ppi_feature.rename({'Unnamed: 0':'Gene'}, axis=1)

#PPI graph construction
ppi_G = nx.from_numpy_matrix(ppi_edges.values)
ppi_G = nx.relabel_nodes(ppi_G, dict(enumerate(ppi_edges.columns)))
ppi_G = nx.relabel_nodes(ppi_G, { n:str(n) for n in ppi_G.nodes()})

ppi_feature.Gene = ppi_feature.Gene.astype(str)
node_attr = ppi_feature.set_index('Gene').to_dict('index')
nx.set_node_attributes(ppi_G, node_attr)

ppi_pyg_graph = from_networkx(ppi_G, group_node_attrs=all)
print(ppi_pyg_graph)

# Data(edge_index=[2, 371078], weight=[371078], x=[1986, 1986])
# There was error when I ran the data.validate(); AttributeError: 'GlobalStorage' object has no attribute 'validate'
```

Then I ran the GCNEncoder code.
I'm sorry to bother you..

**Reply by rusty1s** [reply]
Looks like the issue is that your `x` is of `dtype == torch.long`, while `GCNConv` expects floating point features. Simply doing `x = x.float()` solves this issue.

**Reply by songsong0425** [reply]
It works well! Thank you so much for your help!! 
Can I ask one more question? When I get the embedding as tensor like below, how can I match the value to the corresponding node?
Does it follow the order of nodes in the adjacency matrix?
```
tensor([[-0.0010,  0.0002,  0.0021,  ...,  0.0007, -0.0014,  0.0016], -> which node's embedding?
        [-0.0023,  0.0001,  0.0027,  ...,  0.0020, -0.0018,  0.0023], 
        ...,
        [-0.0318, -0.0364,  0.0301,  ...,  0.0439,  0.0134,  0.0388]], 
       device='cuda:0', grad_fn=<AddBackward0>)
```

**Reply by rusty1s** [reply]
Yes, the order remains unchanged.

**Reply by songsong0425** [reply]
Thanks a lot! I fully understand the process. Have a nice day!
